osu.Game/Online/HubClientConnector.cs

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR.Client;
+using Microsoft.Extensions.DependencyInjection;
+using Newtonsoft.Json;
+using osu.Framework;
+using osu.Framework.Bindables;
+using osu.Framework.Logging;
+using osu.Game.Online.API;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A component that manages the life cycle of a connection to a SignalR Hub.
+    /// </summary>
+    public class HubClientConnector : IDisposable
+    {
+        /// <summary>
+        /// Invoked whenever a new hub connection is built.
+        /// </summary>
+        public Action<HubConnection>? ConfigureConnection;
+
+        private readonly string clientName;
+        private readonly string endpoint;
+        private readonly IAPIProvider? api;
+
+        /// <summary>
+        /// The current connection opened by this connector.
+        /// </summary>
+        public HubConnection? CurrentConnection { get; private set; }
+
+        /// <summary>
+        /// Whether this is connected to the hub, use <see cref="CurrentConnection"/> to access the connection, if this is <c>true</c>.
+        /// </summary>
+        public IBindable<bool> IsConnected => isConnected;
+
+        private readonly Bindable<bool> isConnected = new Bindable<bool>();
+        private readonly SemaphoreSlim connectionLock = new SemaphoreSlim(1);
+        private CancellationTokenSource connectCancelSource = new CancellationTokenSource();
+
+        private readonly IBindable<APIState> apiState = new Bindable<APIState>();
+
+        /// <summary>
+        /// Constructs a new <see cref="HubClientConnector"/>.
+        /// </summary>
+        /// <param name="clientName">The name of the client this connector connects for, used for logging.</param>
+        /// <param name="endpoint">The endpoint to the hub.</param>
+        /// <param name="api"> An API provider used to react to connection state changes, or null to not establish connection at all (for testing purposes).</param>
```

Comment:
i'm still thinking this should be non-nullable and the testing portion should be handled at one level higher if possible. aka you would make a TestMultiplayerClient which overrides the CreateConnector method and returns a null or something.

it feels like a very flimsy parameter in an otherwise solid class.

Reply:
Good point, I've let it override to return a null connector instead.